docs/Architecture Decision Record/006-deployment-flow.md

```diff
@@ -0,0 +1,45 @@
+
+The proposed deploy solution below adds a second Cloud.gov deploy environment to the dev Space. (Each deploy environment is really a pair of applications, one frontend and one backend.) One of the environments will be called `experiment` and used for internal vendor team review. The other will be called `development` and will be used for government review. This aims to solve the "crowding" problem by allowing the vendor team two separate places to deploy code, without giving the vendor team deploy access to two separate Spaces in Cloud.gov.
+
+## Proposed Decision
+
+_Note that this represents a planned future state, not the current state._
+
+Deploy Environment | Cloud.gov Space | Cloud.gov Dev Access | Role                                              | Deploys when ...  |
+-------------------|-----------------|----------------------|---------------------------------------------------|-------------------|
+Experiment         | Dev             | Vendor & Gov         | Deploy code for vendor internal testing & review  | Anytime           |
+Development        | Dev             | Vendor & Gov         | Deploy code submitted for gov review             | Anytime work is ready for QASP review |
+Staging            | Staging         | Gov                  | Deploy code once gov-approved               | Code merged to `HHS/TANF-app:staging`
+Production         | Prod            | Gov                  | Deploy code tested in staging & ready for prod    | Code merged to `HHS/TANF-app:main`
+
+### How would this affect #521? "As a developer, I want a staging site..."
+
+Under the proposed deployment flow, instead of setting up a staging environment in #521, we would set up another pair of frontend/backend applications in the tanf dev Space, which would form the `experiment` deploy environment.
+
+Developers would be able to deploy code to `experiment` anytime. 
+
+The current `development` environment would be used to deploy work ready for QASP review.
+
+We would set up a `staging` environment at a later time -- e.g., once an OFA Tech Lead has onboarded and can own and maintain the staging and deploy sites.
+
+## Consequences
+
+* Good: Vendor dev team "crowding" should be reduced through this solution.
+* Good: Only need to maintain/monitor vendor dev access to a single Cloud.gov Space using this solution.
+* Good: Code deploys automatically upon gov approval, but does not deploy immediately to prod, leaving room for further gov testing.
+* Bad: A four-environment setup has more complexity than a three-environment setup.
+* Alternative: An alternative would be to have a single gov-only Space in Cloud.gov, prod. Gov-approved code would deploy directly to prod on merge. However, this leaves no room for testing code after it has been gov-approved. In this alternative flow, mistakes would deploy immediately to prod.
```

Comment:
@alexsoble This is an interesting, and novel approach. I can see the motivation for suggesting it, though I'm concerned it may be a bit confusing. Conventionally, I would think that the `development` environment would be the one where we can deploy whatever we need to test and experiment with, where the `staging` environment would be there for official reviews and testing. 

I can definitely see how testing code after it has been gov-approved is a good thing, I guess I'm just not sure why this can't be done in the `staging` environment.

Reply:
Hi @carltonsmith, thanks for the feedback! 

Maybe the names I picked for this proposal aren't the clearest — would this naming convention make more sense?

`development` 
`vendor staging`
`gov staging`
`prod`

I can add what the more detailed table would look like if we go this route.

Reply:
Deploy Environment | Cloud.gov Space | Cloud.gov Dev Access | Role                                              | Deploys when ...  |
-------------------|-----------------|----------------------|---------------------------------------------------|-------------------|
Development         | Dev             | Vendor & Gov         | Deploy code for vendor internal testing & review  | Anytime           |
Vendor Staging        | Dev             | Vendor & Gov         | Deploy code submitted for gov review             | Anytime work is ready for QASP review |
Gov Staging            | Staging         | Gov                  | Deploy code once gov-approved               | Code merged to `HHS/TANF-app:staging`
Production         | Prod            | Gov                  | Deploy code tested in staging & ready for prod    | Code merged to `HHS/TANF-app:main`

Reply:
@alexsoble I think this naming convention would cut down greatly on any confusion. I can get on board with this solution :)

Reply:
Updated the proposed ADR with the new naming convention!